examples/reader-writer.cc

```diff
@@ -0,0 +1,426 @@
+constexpr int FIXED_LENGTH = 10;
+const std::string PARQUET_FILENAME = "parquet_cpp_example.parquet";
+
+class ParquetWriter {
+ public:
+  ParquetWriter() {
+      ostream_.reset(new parquet::InMemoryOutputStream());
+  }
+
+  void SetupSchema() {
+    parquet::schema::NodeVector fields;
+    // Create a primitive node named 'boolean_field' with type:BOOLEAN, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("boolean_field", parquet::Repetition::REQUIRED, parquet::Type::BOOLEAN, parquet::LogicalType::NONE));
+    // Create a primitive node named 'int32_field' with type:INT32, repetition:REQUIRED, logical type:TIME_MILLIS
+    fields.push_back(parquet::schema::PrimitiveNode::Make("int32_field", parquet::Repetition::REQUIRED, parquet::Type::INT32, parquet::LogicalType::TIME_MILLIS));
+    // Create a primitive node named 'int64_field' with type:INT64, repetition:REPEATED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("int64_field", parquet::Repetition::REPEATED, parquet::Type::INT64, parquet::LogicalType::NONE));
+    // Create a primitive node named 'int96_field' with type:INT96, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("int96_field", parquet::Repetition::REQUIRED, parquet::Type::INT96, parquet::LogicalType::NONE));
+    // Create a primitive node named 'float_field' with type:FLOAT, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("float_field", parquet::Repetition::REQUIRED, parquet::Type::FLOAT, parquet::LogicalType::NONE));
+    // Create a primitive node named 'double_field' with type:DOUBLE, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("double_field", parquet::Repetition::REQUIRED, parquet::Type::DOUBLE, parquet::LogicalType::NONE));
+    // Create a primitive node named 'ba_field' with type:BYTE_ARRAY, repetition:OPTIONAL
+    fields.push_back(parquet::schema::PrimitiveNode::Make("ba_field", parquet::Repetition::OPTIONAL, parquet::Type::BYTE_ARRAY, parquet::LogicalType::NONE));
+    // Create a primitive node named 'flba_field' with type:FIXED_LEN_BYTE_ARRAY, repetition:REQUIRED
+    fields.push_back(parquet::schema::PrimitiveNode::Make("flba_field", parquet::Repetition::REQUIRED, parquet::Type::FIXED_LEN_BYTE_ARRAY, parquet::LogicalType::NONE, FIXED_LENGTH));
+
+    // Create a GroupNode named 'schema' using the primitive nodes defined above
+    // This GroupNode is the root node of the schema tree
+    schema_ = std::static_pointer_cast<parquet::schema::GroupNode>(parquet::schema::GroupNode::Make("schema", parquet::Repetition::REQUIRED, fields));
+  }
+
+  parquet::ParquetFileWriter* GetWriter() {
+    if (!writer_) {
+      // Create a parquet file writer using the schema and the output stream.
+      writer_ = parquet::ParquetFileWriter::Open(ostream_, schema_);
+    }
+    return writer_.get();
+  }
+
+  void WriteToFile(std::string filename) {
+    std::ofstream parquetFile;
```

Comment:
With writing the file at the end, it builds up the impression that you need to serialise everything in-memory. As mentioned earlier either describe why you do it or add a comment here stating it a bit more or just use the file implementation in Parquet.